Update GHC to 8.8.3 (lts-15.9)

mokev opened the pull request:
The version update is necessary to build Tamarin under Nix as there is no ghc863 target anymore.
Since version 1.0, binary-orphans no longer provides exports for other packages. That functionality has been moved to binary-instances which is now added as an extra dependency.

See also the relevant issue for nixpkgs https://github.com/NixOS/nixpkgs/pull/69356.

rkunnema commented:
Note: This includes #361, so I will merge only after merging #361.

rkunnema commented:
@jdreier Can you say something about moving to lts-14.27? I have no clue what this entails.

jdreier commented:
If Tamarin compiles with the proposed changes and the case studies run correctly, I do not see any problem to update to lts-14.27. We can even move to a more recent version. (Normally we try to move to the latest lts before each release anyway.)

rkunnema commented:
@mokev Would you mind running the case-studies? The last section in https://github.com/tamarin-prover/tamarin-prover/blob/develop/CONTRIBUTING.md explains how to do that.

mokev commented:
I just removed the accidental commits regarding #361. 

I updated GHC to the last available LTS version which required only minimal changes.
The case studies produce the same output as 10159d805f684a9592d91aa01a1ddf5f59764685 (current HEAD of develop).
The proof of `case-studies/ake/dh/UM_three_pass_combined_fixed_analyzed.spthy` seems to have a different order. However, this is also the case for the current head.

jdreier commented:
The reordering of proof steps occurs frequently on this example, so this looks fine.

rkunnema commented:
Cheers @jdreier and @mokev !
